### PR TITLE
[FIX] payment_adyen: not send False information

### PR DESCRIPTION
**Pull request by anko-odoo**
If customer did not have address,email and telephone fields,
"False" would be send to Adyen resulting in these field
containing False values, which in some cases was falsely
triggering anti-fraud system.

task-3159378

### CODE REVIEW

File: addons/payment_adyen/static/src/js/payment_form.js
Comment (Demesmaeker) [inline addons/payment_adyen/static/src/js/payment_form.js:59]: hm... I thought we agreed on not showing the technical errors sent by providers to the customers, only log or send in chatter when necessary?
cc: @dalaOdoo

File: addons/payment_adyen/utils.py
Comment (Demesmaeker) [inline addons/payment_adyen/utils.py:59]:
```suggestion
    stateOrProvince = partner.state_id.code
    if partner.country_id.code not in ["CA", "US", "GB"]:
        stateOrProvince = stateOrProvince or ''
```

**Review by Demesmaeker — COMMENTED** [review]
Hello :duck: 

Thanks for your work here
Just waiting for dala's confirmation, if it's ok, then lgtm ^^

File: addons/payment_adyen/static/src/js/payment_form.js
Comment (anko-odoo) [reply addons/payment_adyen/static/src/js/payment_form.js:59]: I was not aware of that, if @dalaOdoo will confirm this then indeed I will go back to previous version.

Comment (dalaOdoo) [reply addons/payment_adyen/static/src/js/payment_form.js:59]: are we able to know what is the error ?
The customer should know the payment failed but the error message should contain information to understand how to correct. Here he will just check for card number, ... and nothing will change he will stay blocked and frustrated

**Review by Demesmaeker — APPROVED** [review]

File: addons/payment_adyen/utils.py
Comment (AntoineVDV) [inline addons/payment_adyen/utils.py:56]: snake_case

File: addons/payment_adyen/utils.py
Comment (AntoineVDV) [inline addons/payment_adyen/utils.py:57]:
```suggestion
    if partner.country_id.code not in ["CA", "US"]:
```

Comment (AntoineVDV) [reply addons/payment_adyen/utils.py:57]: single-quotes

File: addons/payment_adyen/utils.py
Comment (AntoineVDV) [inline addons/payment_adyen/utils.py:58]:
```suggestion
    if not stateOrProvince and partner.country_id.code in ["CA", "US"]:
        raise
```

File: addons/payment_adyen/utils.py
Comment (AntoineVDV) [inline addons/payment_adyen/utils.py:69]: ValidationError

Comment (AntoineVDV) [inline addons/payment_adyen/utils.py:69]: Translate the message.

Comment (AntoineVDV) [inline addons/payment_adyen/utils.py:69]:
```suggestion
        raise ValueError("Please fill out the address state.")
```

**Review by AntoineVDV — CHANGES_REQUESTED** [review]

File: addons/payment_adyen/utils.py
Comment (anko-odoo) [reply addons/payment_adyen/utils.py:69]: I don't agree, it's another check if other address values are not False.

Comment (anko-odoo) [reply addons/payment_adyen/utils.py:69]: https://docs.adyen.com/risk-management/avs-checks/

File: addons/payment_adyen/utils.py
Comment (AntoineVDV) [inline addons/payment_adyen/utils.py:56]: It's the same but cognitively easier to read: think about it once and then you can just trust the variable name.
```suggestion
    state_or_province = partner.state_id.code or ''
```

File: addons/payment_adyen/utils.py
Comment (AntoineVDV) [inline addons/payment_adyen/utils.py:62]:
```suggestion
        'stateOrProvince': state_or_province,
```

File: addons/payment_adyen/utils.py
Comment (AntoineVDV) [inline addons/payment_adyen/utils.py:69]:
```suggestion
        raise ValidationError(_("Please complete your address details."))
```

File: addons/payment_adyen/utils.py
Comment (AntoineVDV) [reply addons/payment_adyen/utils.py]: Ok but then you're adding a second check, is it intended?

File: addons/payment_adyen/utils.py
Comment (anko-odoo) [inline addons/payment_adyen/utils.py:69]:
```
    if False in address.values() or (partner.country_id.code in ['CA', 'US'] and not state_or_province):
        raise ValidationError(_("Please complete your address details"))
```
Do you think this is acceptable solution? @AntoineVDV

Comment (anko-odoo) [reply addons/payment_adyen/utils.py:69]: Hm, because I can't come up with anything other than this or two checks.

Comment (AntoineVDV) [reply addons/payment_adyen/utils.py:69]: Since the two warnings serve different purposes (prevent empty state details for US and CA and prevent `False` being sent to Adyen for any field), we could get rid of the second one by simply appending `or ''` to all fields.

Comment (anko-odoo) [reply addons/payment_adyen/utils.py:69]: But from what I understand it's not what Adyen expects. https://docs.adyen.com/risk-management/avs-checks/ It says that all values of address are required.

Comment (AntoineVDV) [reply addons/payment_adyen/utils.py:69]: Oh ok I missed that! Then, one plain dumb check that explicitly loops over the dict keys and raises an error. It does not raise when the key is `stateOrProvince` and the country is not in [US, CA, UK] (notice that this pages mentions UK still)

Comment (anko-odoo) [reply addons/payment_adyen/utils.py:69]: Checked, Adyen is refusing if other values are set to ''.

Comment (anko-odoo) [reply addons/payment_adyen/utils.py:69]: I'm not sure if I understand you correctly, is that what you mean?
```suggestion
    for line in address:
        if address[line] == False:
            raise ValidationError(_("Please complete your address details"))
        if line == 'stateOrProvince' and (partner.country_id.code in ['CA', 'US'] and not state_or_province):
            raise ValidationError(_("Please complete your address details"))
```

Comment (AntoineVDV) [reply addons/payment_adyen/utils.py:69]: Yes :)

Use `for key, value in address.items()` though.

File: addons/payment_adyen/utils.py
Comment (AntoineVDV) [inline addons/payment_adyen/utils.py:69]:
```suggestion
            raise ValidationError(_("Please complete your address details."))
```

File: addons/payment_adyen/utils.py
Comment (AntoineVDV) [inline addons/payment_adyen/utils.py:59]:
```suggestion
        'stateOrProvince': partner.state_id.code or '',  # The state is not always required.
```

**Review by AntoineVDV — CHANGES_REQUESTED** [review]

**Comment by AntoineVDV** [comment]
@robodoo delegate+

File: addons/payment_adyen/i18n/payment_adyen.pot
Comment (AntoineVDV) [inline addons/payment_adyen/i18n/payment_adyen.pot:105]:
```suggestion
msgid "Please complete your address details."
```

**Review by AntoineVDV — APPROVED** [review]
@robodoo delegate+

**Comment by anko-odoo** [comment]
@robodoo r+

**Pull request closed without merging**
